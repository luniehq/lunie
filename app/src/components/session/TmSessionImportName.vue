<template>
  <SessionFrame>
    <TmFormStruct :submit="onSubmit">
      <h2 class="session-title">Choose name</h2>
      <div class="session-main bottom-indent">
        <Steps :steps="[`Recover`, `Name`, `Password`]" active-step="Name" />
        <TmFormGroup field-id="import-name" :field-label="fieldLabel">
          <img
            v-if="importedAddress === undefined"
            class="tm-data-msg__icon"
            src="~assets/images/loader.svg"
            alt="a small spinning circle to display loading"
          />
          <div v-else>
            <p class="address">
              {{ importedAddress }}
            </p>
            <!-- only show the retry option if the networks supports more than one algo -->
            <span
              v-if="networkCryptoTypes.length > 1"
              class="retry-link"
              @click="created(true)"
              >{{ incorrectAddressMessage }}</span
            >
          </div>
        </TmFormGroup>
        <TmFormGroup
          :error="$v.$error && $v.name.$invalid"
          field-id="import-name"
          field-label="Account Name"
        >
          <TmField
            id="import-name"
            v-model.trim="name"
            type="text"
            placeholder="Must have at least 3 characters"
            vue-focus="vue-focus"
          />
          <TmFormMsg
            v-if="$v.name.$error && !$v.name.required"
            name="Name"
            type="required"
          />
          <TmFormMsg
            v-if="$v.name.$error && !$v.name.minLength"
            name="Name"
            type="minLength"
            min="3"
          />
          <TmFormMsg
            v-if="$v.name.$error && !$v.name.nameExists"
            name="Name"
            type="custom"
            msg="already exists"
          />
        </TmFormGroup>
      </div>
      <div class="session-footer">
        <TmBtn value="Next" type="submit" />
      </div>
    </TmFormStruct>
  </SessionFrame>
</template>

<script>
import { required, minLength } from "vuelidate/lib/validators"
import TmBtn from "common/TmBtn"
import TmFormGroup from "common/TmFormGroup"
import TmFormStruct from "common/TmFormStruct"
import TmField from "common/TmField"
import TmFormMsg from "common/TmFormMsg"
import SessionFrame from "common/SessionFrame"
import { mapGetters } from "vuex"
import Steps from "../../ActionModal/components/Steps"
import { getWalletIndex } from "@lunie/cosmos-keys"
import { getWallet } from "../../vuex/modules/wallet"

const nameExists = (value) => {
  const walletIndex = getWalletIndex()
  if (walletIndex.some((e) => e.name === value)) {
    return false
  } else {
    return true
  }
}

export default {
  name: `session-import-name`,
  components: {
    TmBtn,
    TmField,
    SessionFrame,
    TmFormGroup,
    TmFormMsg,
    TmFormStruct,
    Steps,
  },
  data: () => ({
    importedAddress: undefined,
    displayCryptoView: false,
    attempt: 0,
  }),
  computed: {
    ...mapGetters([`network`, `currentNetwork`]),
    name: {
      get() {
        return this.$store.state.recover.name
      },
      set(value) {
        this.$store.commit(`updateField`, { field: `name`, value })
      },
    },
    networkCryptoTypes() {
      if (this.currentNetwork.network_type === `cosmos`) {
        return JSON.parse(this.currentNetwork.HDPaths)
      } else if (this.currentNetwork.network_type === `polkadot`) {
        return JSON.parse(this.currentNetwork.curves)
      } else {
        return []
      }
    },
    currentCryptoView() {
      if (this.currentNetwork.network_type === `cosmos`) {
        return JSON.parse(this.currentNetwork.HDPaths)[this.attempt].name
      } else if (this.currentNetwork.network_type === `polkadot`) {
        return JSON.parse(this.currentNetwork.curves)[this.attempt].name
      } else {
        return ``
      }
    },
    incorrectAddressMessage() {
      const messages = [
        `Not your address?`,
        `Maybe this one?`,
        `Still not your address?`,
      ]
      return messages[this.attempt]
    },
    fieldLabel() {
      if (this.displayCryptoView) {
        return `Your Address created using ` + this.currentCryptoView
      } else {
        return `Your Address`
      }
    },
  },
  mounted() {
    this.created()
  },
  methods: {
    onSubmit() {
      this.$v.$touch()
      if (this.$v.name.$invalid) return
      this.$router.push("/recover/password")
    },
    currentHDPathOrCurve() {
      if (this.currentNetwork.network_type === `cosmos`) {
        return {
          HDPath: this.networkCryptoTypes[this.attempt].value,
          curve: JSON.parse(this.currentNetwork.curves)[0].value, // ed25519
        }
      } else if (this.currentNetwork.network_type === `polkadot`) {
        return {
          HDPath: JSON.parse(this.currentNetwork.HDPaths)[0].value, // no clue
          curve: this.networkCryptoTypes[this.attempt].value,
        }
      } else {
        return {
          HDPath: ``,
          curve: ``,
        }
      }
    },
    async created(retry = false) {
      if (retry) {
        this.attempt++
        this.displayCryptoView = true
      }
      this.attempt = this.numberAttemptsController(
        this.networkCryptoTypes,
        this.attempt
      )
      const wallet = await getWallet(
        this.$store.state.recover.seed,
        this.currentNetwork,
        this.currentHDPathOrCurve()[`HDPath`],
        this.currentHDPathOrCurve()[`curve`]
      )
      this.importedAddress = wallet.cosmosAddress
      this.$store.commit(`updateField`, {
        field: `HDPath`,
        value: this.currentHDPathOrCurve()[`HDPath`],
      })
      this.$store.commit(`updateField`, {
        field: `curve`,
        value: this.currentHDPathOrCurve()[`curve`],
      })
    },
    numberAttemptsController(networkCryptoTypes, attempt) {
      if (attempt >= networkCryptoTypes.length) {
        return attempt - networkCryptoTypes.length
      } else {
        return attempt
      }
    },
  },
  validations: () => ({
    name: { required, minLength: minLength(3), nameExists },
  }),
}
</script>
<style scoped>
.address {
  word-break: break-all;
  font-size: 0.9rem;
  color: var(--txt);
}

.retry-link {
  font-size: 12px;
  cursor: pointer;
  color: var(--link);
}

.retry-link:hover {
  color: var(--link-hover);
}

.tm-form-group__field {
  position: unset !important;
}
</style>
