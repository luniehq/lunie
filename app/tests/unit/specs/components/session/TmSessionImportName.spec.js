import Vuex from "vuex"
import Vuelidate from "vuelidate"
import { shallowMount, createLocalVue } from "@vue/test-utils"
import TmSessionImportName from "session/TmSessionImportName"
jest.mock(`scripts/google-analytics.js`, () => () => {})
jest.mock("@lunie/cosmos-keys", () => ({
  getWalletIndex: function () {
    return [{ name: `Happy Lunie User`, address: `xyz123` }]
  },
  getNewWalletFromSeed: function () {
    return { cosmosAddress: `cosmos1234` }
  },
}))
const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(Vuelidate)
localVue.directive(`tooltip`, () => {})
localVue.directive(`focus`, () => {})

describe(`TmSessionImportName`, () => {
  let wrapper, $store, getters

  beforeEach(() => {
    getters = {
      connected: () => true,
      network: "cosmos-hub-mainnet",
      currentNetwork: {
        id: "cosmos-hub-mainnet",
        network_type: "cosmos",
        address_prefix: "cosmos",
        testnet: false,
        HDPaths: `[{"value":"m/44'/118'/0'/0/0", "name":"Cosmos HD Path"}]`,
        curves: `[{"value":"ed25519", "name":"Edwards curve"}]`,
      },
    }
    $store = {
      state: {
        recover: {
          name: ``,
          seed: ``,
          password: ``,
          passwordConfirm: ``,
          warning: false,
        },
      },
      getters,
      commit: jest.fn(),
      dispatch: jest.fn(async () => `cosmos1234`),
      mutations: {
        updateField: jest.fn(),
      },
    }
    wrapper = shallowMount(TmSessionImportName, {
      localVue,
      mocks: {
        $store,
        $router: {
          push: jest.fn(),
        },
      },
      stubs: [`router-link`],
    })
  })

  it(`has the expected html structure`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`validation should fail if name is not filled in`, async () => {
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.name.$error).toBe(true)
  })

  it(`validation should fail if name lenght < 3 characters`, async () => {
    wrapper.vm.$store.state.recover.name = `as`
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.name.$error).toBe(true)
  })

  it(`validation should not fail if name lenght >= 3 characters`, async () => {
    wrapper.vm.$store.state.recover.name = `Happy Lunie User 2`
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.name.$error).toBe(false)
  })

  it(`validation should fail if name exists already in stored accounts`, async () => {
    wrapper.vm.$store.state.recover.name = `Happy Lunie User`
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.name.$error).toBe(true)
  })

  it(`should commit updateField on field change`, async () => {
    wrapper.setData({ name: `Happy Lunie User` })
    expect($store.commit).toHaveBeenCalledWith(`updateField`, {
      field: `name`,
      value: `Happy Lunie User`,
    })
  })

  it(`should go to /recover/password when submit the form`, async () => {
    wrapper.vm.$store.state.recover.name = `Happy Lunie User 2`
    wrapper.vm.onSubmit()
    expect(wrapper.vm.$router.push).toHaveBeenCalledWith(`/recover/password`)
  })

  // case cosmos
  it(`should return the HDPaths from a network parsed as JSON`, () => {
    expect(wrapper.vm.networkCryptoTypes).toEqual([
      { value: "m/44'/118'/0'/0/0", name: "Cosmos HD Path" },
    ])
  })

  // case polkadot
  it(`should return the curves from a network parsed as JSON`, () => {
    wrapper.setData({
      currentNetwork: {
        id: "polkadot-testnet",
        network_type: "polkadot",
        address_prefix: "42",
        testnet: false,
        HDPaths: `[""]`,
        curves: `[{"value":"ed25519", "name":"Edwards curve"}]`,
      },
    })
    expect(wrapper.vm.networkCryptoTypes).toEqual([
      { value: "ed25519", name: "Edwards curve" },
    ])
  })

  // case cosmos
  it(`should return the current HDPath we are creating the address from depending on the attempt number`, () => {
    wrapper.setData({
      attempt: 0,
    })
    const curve = wrapper.vm.currentHDPathOrCurve()["HDPath"]
    expect(curve).toEqual("m/44'/118'/0'/0/0")
  })

  // case polkadot
  it(`should return the current curve we are creating the address from depending on the attempt number`, () => {
    wrapper.setData({
      currentNetwork: {
        id: "polkadot-testnet",
        network_type: "polkadot",
        address_prefix: "42",
        testnet: false,
        HDPaths: `[""]`,
        curves: `[{"value":"sr25519", "name":"Schnorrkel curve"},{"value":"ed25519", "name":"Edwards curve"},{"value":"ecdsa", "name":"ECDSA"}]`,
      },
      attempt: 1,
    })
    const curve = wrapper.vm.currentHDPathOrCurve()["curve"]
    expect(curve).toEqual("ed25519")
  })

  // case cosmos
  it(`should return a beautiful presentation for the curve to be displayed on modal`, () => {
    expect(wrapper.vm.currentCryptoView).toEqual("Cosmos HD Path")
  })

  // case polkadot
  it(`should return a beautiful presentation for the curve to be displayed on modal`, () => {
    wrapper.setData({
      currentNetwork: {
        id: "polkadot-testnet",
        network_type: "polkadot",
        address_prefix: "42",
        testnet: false,
        HDPaths: `[""]`,
        curves: `[{"value":"sr25519", "name":"Schnorrkel curve"},{"value":"ed25519", "name":"Edwards curve"}]`,
      },
      attempt: 1,
    })
    expect(wrapper.vm.currentCryptoView).toEqual("Edwards curve")
  })
})
