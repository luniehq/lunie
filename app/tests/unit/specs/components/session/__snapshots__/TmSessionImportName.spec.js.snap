// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmSessionImportName has the expected html structure 1`] = `
<sessionframe-stub
  icon=""
>
  <tmformstruct-stub
    submit="function () { [native code] }"
    width=""
  >
    <h2
      class="session-title"
    >
      Choose name
    </h2>
     
    <div
      class="session-main bottom-indent"
    >
      <steps-stub
        activestep="Name"
        steps="Recover,Name,Password"
      />
       
      <tmformgroup-stub
        fieldid="import-name"
        fieldlabel="Your Address"
      >
        <img
          alt="a small spinning circle to display loading"
          class="tm-data-msg__icon"
          src="~assets/images/loader.svg"
        />
      </tmformgroup-stub>
       
      <tmformgroup-stub
        fieldid="import-name"
        fieldlabel="Account Name"
      >
        <tmfield-stub
          id="import-name"
          placeholder="Must have at least 3 characters"
          type="text"
          value=""
          vue-focus="vue-focus"
        />
         
        <!---->
         
        <!---->
         
        <!---->
      </tmformgroup-stub>
    </div>
     
    <div
      class="session-footer"
    >
      <tmbtn-stub
        customclass=""
        type="submit"
        value="Next"
      />
    </div>
  </tmformstruct-stub>
</sessionframe-stub>
`;
